File: assignment2/src/assignment2.c
#include <netinet/ether.h>
#include <net/ethernet.h>
#include <stdio.h>
#include <arpa/inet.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <stdlib.h>
#include <netinet/ip6.h>
#include <netinet/icmp6.h>
#include <asm/byteorder.h>

#include "arguments.h"
#include "raw.h"
#include "hexdump.h"
#include "checksums.h"

/*====================================TODO===================================*/
/* Put your required struct definitions */

uint8_t mymac[ETH_ALEN];

/* Put your ancillary functions here*/

/*===========================================================================*/

void assignment2(int fd, int frames)
{
	unsigned int timeout = 10000;
	uint8_t recbuffer[1514];
	size_t ret;

/*====================================TODO===================================*/
	/* If you want to set up any data/counters before the receive loop,
	 * this is the right location
	 */

	memcpy(&mymac, grnvs_get_hwaddr(fd), ETH_ALEN);
	int allEtherBytes[65535];
	int allEtherFrames[65535];
	
	int multicast = 0;
	int notMulticast = 0;
	int allBytes = 0;
/*===========================================================================*/

	/* This is the ready marker! do not remove! */
	fprintf(stdout, "I am ready!\n");

/*====================================TODO===================================*/
	/* Update the loop condition */
	while(frames-- > 0) {
/*===========================================================================*/
		ret = grnvs_read(fd, recbuffer, sizeof(recbuffer), &timeout);
		if (ret == 0) {
			fprintf(stderr, "Timed out, this means there was nothing to receive. Do you have a sender set up?\n");
			break;
		}
/*====================================TODO===================================*/
	/* This is the receive loop, 'recbuffer' will contain the received
	 * frame. 'ret' tells you the length of what you received.
	 * Anything that should be done with every frame that's received
	 * should be done here.
	 */
		uint8_t etherTypeF = recbuffer[12];
		uint8_t etherTypeB = recbuffer[13];
		int index = etherTypeF << 8 | etherTypeB;
		allEtherBytes[index] += ret;
		allEtherFrames[index]++;
		allBytes += ret;
		if((recbuffer[0] & (1<<7)) == 1)//recbuffer[0] & 1000 0000
		{
			multicast++;
		}
		else 
		{
			notMulticast++;
		}
/*===========================================================================*/
	}

/*====================================TODO===================================*/
	/* Print your summary here */
	if(allEtherFrames[2048] > 0)
	{
		printf("0x0800: %d frames, %d bytes\n", allEtherFrames[2048], allEtherBytes[2048]);
	}
	if(allEtherFrames[2054] > 0){

		printf("0x0806: %d frames, %d bytes\n", allEtherFrames[2054], allEtherBytes[2054]);
	}
	if(allEtherFrames[34525] > 0){

		printf("0x86dd: %d frames, %d bytes\n", allEtherFrames[34525], allEtherBytes[34525]);
	}
	printf("%d of them were for me\n", notMulticast);
	printf("%d of them were multicast\n", multicast);
	if(allBytes > 0)
	{
		printf("IPv4 accounted for %f%% and IPv6 for %f%% of traffic\n", 100.0 * allEtherBytes[2048]/allBytes, 100.0 * allEtherBytes[34525]/allBytes);
	}
/*===========================================================================*/
}

int main(int argc, char ** argv)
{
	struct arguments args;
	int sock;

	setvbuf(stdout, NULL, _IOLBF, 0);

	if ( parse_args(&args, argc, argv) < 0 ) {
		fprintf(stderr, "Failed to parse arguments, call with "
			"--help for more information\n");
		return -1;
	}

	if ( (sock = grnvs_open(args.interface, SOCK_RAW)) < 0 ) {
		fprintf(stderr, "grnvs_open() failed: %s\n", strerror(errno));
		return -1;
	}

	assignment2(sock, args.frames);

	grnvs_close(sock);

	return 0;
}

File: assignment2/src/Assignment2.java
/* DO NOT!!! put a package statement here, that would break the build system */
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.HashMap;

public class Assignment2 {

/*====================================TODO===================================*/
	/* Put your required state variables here */
	static byte[] mymac;

/*===========================================================================*/

	public static void run(GRNVS_RAW sock, int frames) {
		byte[] recbuffer = new byte[1514];
		int length;
		Timeout timeout = new Timeout(10000);

/*====================================TODO===================================*/
	/* If you want to set up any data/counters before the receive loop,
	 * this is the right location
	 */
		mymac = sock.getMac();
		int end = frames;
		int multicast = 0;
		int notMulticast = 0;
		HashMap<int[],int[]> allEtherTypes = new HashMap<>();
/*===========================================================================*/

	/* keep this AS IS! the tester uses this to avoid races */
		System.out.format("I am ready!\n");

/*====================================TODO===================================*/
	/* Update loop condition */
		while (end-- > 0) {
/*===========================================================================*/
			length = sock.read(recbuffer, timeout);
			if (length == 0) {
				System.err.format("Timed out, this means there was nothing to receive. Do you have a sender set up?\n");
				break;
			}
/*====================================TODO===================================*/
	/* This is the receive loop, 'recbuffer' will contain the received
	 * frame. 'length' tells you the length of what you received.
	 * Anything that should be done with every frame that's received
	 * should be done here.
	 */
			int[] addTo = allEtherTypes.get(new int[]{recbuffer[12],recbuffer[13]});
			if(addTo != null)
			{
				allEtherTypes.put(new int[]{recbuffer[12],recbuffer[13]},new int[]{addTo[0] + 1,addTo[1] + 1});
			}
			else
			{
				allEtherTypes.put(new int[]{recbuffer[12],recbuffer[13]},new int[]{1, 1});
			}
			if(recbuffer[0] > 127){

				notMulticast++;
			}
			else{

				multicast++;
			}
/*===========================================================================*/
		}
/*====================================TODO===================================*/
	/* Print your summary here */
		if (allEtherTypes.get(new int[]{8,0}) == null) {
			allEtherTypes.put(new int[]{8,0},new int[]{0,0});
		}
		if (allEtherTypes.get(new int[]{8,6}) == null) {
			allEtherTypes.put(new int[]{8,6},new int[]{0,0});
		}
		if (allEtherTypes.get(new int[]{134,221}) == null) {
			allEtherTypes.put(new int[]{134,221},new int[]{0,0});
		}
		if(allEtherTypes.get(new int[]{8,0})[0] > 0){

			System.out.printf("0x0800: %d frames, %d bytes\n", allEtherTypes.get(new int[]{8,0})[0], allEtherTypes.get(new int[]{8,0})[1]);
		}
		if(allEtherTypes.get(new int[]{8,6})[0] > 0){

			System.out.printf("0x0806: %d frames, %d bytes\n", allEtherTypes.get(new int[]{8,6})[0], allEtherTypes.get(new int[]{8,6})[1]);
		}
		if(allEtherTypes.get(new int[]{134,221})[0] > 0){

			System.out.printf("0x86dd: %d frames, %d bytes\n", allEtherTypes.get(new int[]{134,221})[0], allEtherTypes.get(new int[]{134,221})[1]);
		}
		System.out.printf("%d of them were for me\n", notMulticast);
		System.out.printf("%d of them were multicast\n", multicast);
		int allBytes = allEtherTypes.get(new int[]{8,0})[1] + allEtherTypes.get(new int[]{8,6})[1] + allEtherTypes.get(new int[]{134,221})[1];
		if(allBytes > 0){
			System.out.printf("IPv4 accounted for %f%% and IPv6 for %f%% of traffic\n", 100.0 * allEtherTypes.get(new int[]{8,0})[1]/allBytes, 100.0 * allEtherTypes.get(new int[]{134,221})[1]/allBytes);
		}
/*===========================================================================*/
	}


	public static void main(String[] argv) {
		Arguments args = new Arguments(argv);
		GRNVS_RAW sock = null;
		try{
			sock = new GRNVS_RAW(args.iface, 3);
			run(sock, args.frames);
			sock.close();
		}
		catch(Exception e) {
			e.printStackTrace();
			System.out.println(e.getMessage());
			System.exit(1);
		}
	}
}
